File: src/component/Component.java
package component;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import org.newdawn.slick.Color;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.geom.Rectangle;
import org.newdawn.slick.geom.Shape;
import org.newdawn.slick.geom.Vector2f;
import org.newdawn.slick.gui.AbstractComponent;
import org.newdawn.slick.gui.GUIContext;

//import component.Positionable.ReferencePoint;

import scene.Scene;

/**
 * {@code Component} provides basic features for GUI elements with an origin,
 * width, and height.
 */
public abstract class Component extends AbstractComponent implements Positionable, Visible {
	public static enum DebugMode {
		NONE,
		VISIBLE,
		ALL
	}
	private static DebugMode debugMode;
	
	public static enum BevelType {
		NONE,
		IN,
		OUT
	}
	
	private Visible visibleParent;
	private Vector2f origin;
	private Vector2f translation;
	private int width;
	private int height;
	
	private Color backgroundColor;
	private BevelType beveled;
	private int bevelWidth;
	
	private Color borderColor;
	private int topBorderWidth;
	private int rightBorderWidth;
	private int bottomBorderWidth;
	private int leftBorderWidth;
	
	protected List<Component> components;
	private boolean mouseOver;
	private boolean visible;
	
	private boolean tooltipEnabled;
	private String tooltipMessage;
	private boolean shouldUpdateComponents;
	
	/**
	 * Constructs a {@code Component} with a width and height.
	 * @param context The Gui Context
	 * @param width the width of the component
	 * @param height the height of the component
	 */
	public Component(GUIContext context, int width, int height) {
		super(context);
		
		origin = new Vector2f();
		translation = new Vector2f();
		this.width = width;
		this.height = height;
		
		components = new ArrayList<Component>();
		visible = true;
	}
	
	@Override
	public void render(GUIContext context, Graphics g) throws SlickException {
		g.setClip((Rectangle) getArea());
		
		if (backgroundColor != null) {
			g.setColor(backgroundColor);
			if (beveled == BevelType.NONE) {
				g.fillRect(getX(), getY(), getWidth(), getHeight());
			} else {
				Color brightColor = backgroundColor.brighter(0.2f);
				Color darkColor = backgroundColor.darker(0.2f);
				
				g.setColor(backgroundColor);
				// inner rect
				g.fillRect(getX() + bevelWidth,
						getY() + bevelWidth,
						getWidth() - bevelWidth * 2,
						getHeight() - bevelWidth * 2);
				
				// top bar
				if (beveled == BevelType.OUT) {
					g.setColor(brightColor);
				} else if (beveled == BevelType.IN) {
					g.setColor(darkColor);
				}
				g.fillRect(getX() + leftBorderWidth,
						getY() + topBorderWidth,
						getWidth() - leftBorderWidth - rightBorderWidth,
						bevelWidth);
						
				// bottom bar
				if (beveled == BevelType.OUT) {
					g.setColor(darkColor);
				} else if (beveled == BevelType.IN) {
					g.setColor(brightColor);
				}
				g.fillRect(getX() + leftBorderWidth,
						getY() + getHeight() - bevelWidth - bottomBorderWidth,
						getWidth() - leftBorderWidth - rightBorderWidth,
						bevelWidth);
						
				// left bar
				if (beveled == BevelType.OUT) {
					g.setColor(brightColor);
				} else if (beveled == BevelType.IN) {
					g.setColor(darkColor);
				}
				g.fillRect(getX() + leftBorderWidth,
						getY() + topBorderWidth,
						bevelWidth,
						getHeight() - topBorderWidth - bottomBorderWidth);
				
				// right bar
				if (beveled == BevelType.OUT) {
					g.setColor(darkColor);
				} else if (beveled == BevelType.IN) {
					g.setColor(brightColor);
				}
				g.fillRect(getX() + getWidth() - bevelWidth - rightBorderWidth,
						getY() + topBorderWidth,
						bevelWidth,
						getHeight() - topBorderWidth - bottomBorderWidth);
			}
		}
		
		for (Component component : components) {
			component.render(container, g);
			
			if (debugMode == DebugMode.ALL || debugMode == DebugMode.VISIBLE && component.isVisible()) {
				if (component.isVisible()) {
					g.setColor(Color.red);
				} else {
					g.setColor(Color.yellow);
				}
				g.drawRect(component.getX(), component.getY(), component.getWidth() - 1, component.getHeight() - 1);
			}
		}
		
		g.setClip((Rectangle) getArea());
				
		// border
		g.setColor(borderColor);
		// top bar
		g.fillRect(getX(),
				getY(),
				getWidth(),
				topBorderWidth);
		
		// right bar
		g.fillRect(getX() + getWidth() - rightBorderWidth,
				getY(),
				rightBorderWidth,
				getHeight());
		
		// bottom bar
		g.fillRect(getX(),
				getY() + getHeight() - bottomBorderWidth,
				getWidth(),
				bottomBorderWidth);
				
		// left bar
		g.fillRect(getX(),
				getY(),
				leftBorderWidth,
				getHeight());
		
		g.clearClip();
	}
	
	/**
	 * Update the component based on the time since last update.
	 * @param delta Time since last update in milliseconds.
	 */
	public void update(int delta) {
		if (shouldUpdateComponents()) {
			for (Component c : components) {
				c.update(delta);
			}
		}
		return;
	}
	
	public void setShouldUpdateComponents(boolean shouldUpdate) {
		shouldUpdateComponents = shouldUpdate;
	}
	
	private boolean shouldUpdateComponents() {
		return shouldUpdateComponents;
	}

	/**
	 * Sets the {@code Visible} parent.
	 * @param visible New visible parent
	 */
	public void setVisibleParent(Visible visible) {
		this.visibleParent = visible;
	}
	
	/**
	 * Return if visible and visible parent (and grandparents) is/are visible.
	 * @return {@code true} if visible, {@code false} is not
	 */
	@Override
	public boolean isVisible() {
		if (visibleParent != null) {
			return visible && visibleParent.isVisible();
		}
		
		return false;
	}
	
	/**
	 * Set this component to be visible or invisible.
	 * @param visible Enables visibility if {@code true}, disables if {@code false}
	 */
	@Override
	public void setVisible(boolean visible) {
		this.visible = visible;
	}
	
	/**
	 * Add this component at a location relative to a reference point.
	 * @param component {@code Component} to add
	 * @param location Location to add the component at
	 * @param referencePoint ReferencePoint in the component to set to the location
	 * @param xOffset X Offset from the location
	 * @param yOffset Y Offset from the location
	 */
	public void add(Component component, Vector2f location, ReferencePoint referencePoint, int xOffset, int yOffset) {
		component.setPosition(location, referencePoint, xOffset, yOffset);
		
		components.add(component);
		component.setVisibleParent(this);
	}
	
	/**
	 * Add this component at a location relative to a reference point.
	 * @param component {@code Component} to add
	 * @param location Location to add the component at
	 * @param referencePoint {@code ReferencePoint} in the component to set to the location
	 */
	public void add(Component component, Vector2f location, ReferencePoint referencePoint) {
		add(component, location, referencePoint, 0, 0);
	}
	
	/**
	 * Add a set of components in a column with given spacing.
	 * @param components {@code Component}s to add
	 * @param location Location to add the components column at
	 * @param xOffset X Offset from the location
	 * @param yOffset Y Offset from the location
	 * @param spacing Spacing between components in the column
	 */
	public void addAsColumn(Iterator<? extends Component> components, Vector2f location, int xOffset, int yOffset, int spacing) {
		while (components.hasNext()) {
			Component  c = components.next();
			add(c, location, ReferencePoint.TOPLEFT, xOffset, yOffset);
			location.set(location.x, location.y + c.getHeight() + spacing);
		}
	}
	
	/**
	 * Add a set of components in a row with given spacing.
	 * @param components {@code Component}s to add
	 * @param location Location to add the components row at
	 * @param xOffset X Offset from the location
	 * @param yOffset Y Offset from the location
	 * @param spacing Spacing between components in the row
	 */
	public void addAsRow(Iterator<? extends Component> components, Vector2f location, int xOffset, int yOffset, int spacing) {
		while (components.hasNext()) {
			Component c = components.next();
			if (c != null) {
				add(c, location, ReferencePoint.TOPLEFT, xOffset, yOffset);
				location.set(location.x + c.getWidth() + spacing, location.y);
			}
		}
	}
	
	/**
	 * Add a set of components in a row with given spacing.
	 * @param components {@code Component}s to add
	 * @param location Location to add the components row at
	 * @param rows Number of rows to have
	 * @param cols Number of cols to have
	 * @param xOffset X Offset from the location
	 * @param yOffset Y Offset from the location
	 * @param xSpacing X Spacing between components in the row
	 * @param ySpacing Y Spacing between components in the row
	 */
	public void addAsGrid(Component[] components, Vector2f location, int rows, int cols, int xOffset, int yOffset, int xSpacing, int ySpacing) {
		int startIndex = 0;
		
		while (startIndex < components.length) {
			Component[] row = new Component[cols];
			for (int x = 0; x < row.length && startIndex + x < components.length; x++) {
				row[x] = components[startIndex + x];
			}
			
			addAsRow(Arrays.asList(row).iterator(), location.copy(), xOffset, yOffset, xSpacing);
			
			location.set(location.x, location.y + components[startIndex].getHeight() + ySpacing);
			startIndex += cols;
		}
	}
	
	/**
	 * Add a set of components in a row with given spacing.
	 * @param components {@code Component}s to add
	 * @param location Location to add the components row at
	 * @param rows Number of rows to have
	 * @param cols Number of cols to have
	 * @param xOffset X Offset from the location
	 * @param yOffset Y Offset from the location
	 * @param xSpacing X Spacing between components in the row
	 * @param ySpacing Y Spacing between components in the row
	 */
	public void addAsGrid(Component[][] components, Vector2f location, int xSpacing, int ySpacing) {
		for (int r = 0; r < components.length; r++) {
			addAsRow(Arrays.asList(components[r]).iterator(), location.copy(), 0, 0, xSpacing);
			
			location.set(location.x, location.y + components[r][0].getHeight() + ySpacing);
		}
	}
	
	/**
	 * Remove a {@code Component}.
	 * @param component Component to remove
	 */
	public void remove(Component component) {
		component.setVisibleParent(null);
		components.remove(component);
	}
	
	@Override
	public void setPosition(Vector2f location, ReferencePoint referencePoint) {
		int x = (int) location.x;
		int y = (int) location.y;
		
		switch (referencePoint) {
			case TOPLEFT:
				setLocation(x, y);
				break;
			case TOPCENTER:
				setLocation(x - getWidth() / 2, y);
				break;
			case TOPRIGHT:
				setLocation(x - getWidth(), y);
				break;
			case CENTERLEFT:
				setLocation(x, y - getHeight() / 2);
				break;
			case CENTERCENTER:
				setLocation(x - getWidth() / 2, y - getHeight() / 2);
				break;
			case CENTERRIGHT:
				setLocation(x - getWidth(), y - getHeight() / 2);
				break;
			case BOTTOMLEFT:
				setLocation(x, y - getHeight());
				break;
			case BOTTOMCENTER:
				setLocation(x - getWidth() / 2, y - getHeight());
				break;
			case BOTTOMRIGHT:
				setLocation(x - getWidth(), y - getHeight());
				break;
		}
	}
	
	@Override
	public void setPosition(Vector2f location, ReferencePoint referencePoint, int xOffset, int yOffset) {
		Vector2f offsetLocation = new Vector2f(location.x + xOffset, location.y + yOffset);
		
		setPosition(offsetLocation, referencePoint);
	}
	
	@Override
	public Vector2f getPosition(ReferencePoint referencePoint) {
		int x = getX();
		int y = getY();
		
		switch (referencePoint) {
			case TOPLEFT:
				return new Vector2f(x, y);
			case TOPCENTER:
				return new Vector2f(x + getWidth() / 2, y);
			case TOPRIGHT:
				return new Vector2f(x + getWidth(), y);
			case CENTERLEFT:
				return new Vector2f(x, y + getHeight() / 2);
			case CENTERCENTER:
				return new Vector2f(x + getWidth() / 2, y + getHeight() / 2);
			case CENTERRIGHT:
				return new Vector2f(x + getWidth(), y + getHeight() / 2);
			case BOTTOMLEFT:
				return new Vector2f(x, y + getHeight());
			case BOTTOMCENTER:
				return new Vector2f(x + getWidth() / 2, y + getHeight());
			case BOTTOMRIGHT:
				return new Vector2f(x + getWidth(), y + getHeight());
		}
		
		return null;
	}
	
	/**
	 * Get the background color.
	 * @return The background color
	 */
	public Color getBackgroundColor() {
		return backgroundColor;
	}
	
	/**
	 * Set the background color.
	 * @param color The new background color
	 */
	public void setBackgroundColor(Color color) {
		this.backgroundColor = color;
	}
	
	/**
	 * Sets a beveled appearance.
	 * @param bevelType The {@code BevelType} to use
	 */
	public void setBevel(BevelType bevelType) {
		this.beveled = bevelType;
	}
	
	/**
	 * Set the bevel width.
	 * @param width The new bevel width
	 */
	public void setBevelWidth(int width) {
		this.bevelWidth = width;
	}
	
	/**
	 * Set the border width of all sides.
	 * @param width The new border width of all sides
	 */
	public void setBorderWidth(int width) {
		setTopBorderWidth(width);
		setRightBorderWidth(width);
		setBottomBorderWidth(width);
		setLeftBorderWidth(width);
	}
	
	/**
	 * Set the bottom border width.
	 * @param width The new top border width
	 */
	public void setTopBorderWidth(int width) {
		this.topBorderWidth = width;
	}
	
	/**
	 * Set the right border width.
	 * @param width The new right border width
	 */
	public void setRightBorderWidth(int width) {
		this.rightBorderWidth = width;
	}
	
	/**
	 * Set the bottom border width.
	 * @param width The new bottom border width
	 */
	public void setBottomBorderWidth(int width) {
		this.bottomBorderWidth = width;
	}
	
	/**
	 * Set the left border width.
	 * @param width The new left border width
	 */
	public void setLeftBorderWidth(int width) {
		this.leftBorderWidth = width;
	}
	
	/**
	 * Set the border color.
	 * @param color The new color
	 */
	public void setBorderColor(Color color) {
		this.borderColor = color;
	}
	
	/**
	 * Return this area with origin, width, and height.
	 * @return This area's origin, width, and height
	 */
	public final Shape getArea() {
		return new Rectangle(origin.x, origin.y, this.width, this.height);
	}
	
	@Override
	public final int getWidth() {
		return width;
	}
	
	@Override
	public final int getHeight() {
		return height;
	}
	
	@Override
	public final int getX() {
		return (int) (origin.getX() + translation.getX());
	}

	@Override
	public final int getY() {
		return (int) (origin.getY() + translation.getY());
	}
	
	@Override
	public void setLocation(int x, int y) {
		if (origin == null) {
			origin = new Vector2f(x, y);
		} else {
			if (components != null) {
				for (Component c : components) {
					c.setLocation(c.getX() + (x - getX()), c.getY() + (y - getY()));
				}
			}
			origin.set(x, y);
		}
	}
	
	public void setTranslation(int x, int y) {
		if (components != null) {
			for (Component c : components) {
				c.setTranslation(x, y);
			}
		}
		translation.set(x, y);
	}
	
	@Override
	public void setAcceptingInput(boolean acceptingInput) {
		super.setAcceptingInput(acceptingInput);
		
		if (!acceptingInput) {
			mouseOver = false;
		}
		
		for (Component c : components) {
			c.setAcceptingInput(acceptingInput);
		}
	}
	
	/**
	 * Returns the {@code Visible} parent.
	 * @return Return the visible parent
	 */
	public Visible getVisibleParent() {
		return visibleParent;
	}
	
	@Override
	public void mouseMoved(int oldx, int oldy, int newx, int newy) {
		if (!isVisible() || !isAcceptingInput()) {
			return;
		}
		
		super.mouseMoved(oldx, oldy, newx, newy);
		
		this.mouseOver = getArea().contains(newx, newy);
		
		if (isMouseOver() && tooltipEnabled) {
			Scene.showTooltip(newx, newy, this, tooltipMessage);
		}
	}
	
	@Override
	public void mousePressed(int button, int mx, int my) {
		if (!isVisible() || !isAcceptingInput()) {
			return;
		}
		
		super.mousePressed(button, mx, my);
		
		this.mouseOver = getArea().contains(mx, my);
	}
	
	@Override
	public void mouseReleased(int button, int mx, int my) {
		if (!isVisible() || !isAcceptingInput()) {
			return;
		}
		
		super.mouseReleased(button, mx, my);
		
		this.mouseOver = getArea().contains(mx, my);
	}
	
	/**
	 * Return if mouse is over.
	 * @return {@code true} if mouse is over, {@code false} is not
	 */
	public boolean isMouseOver() {
		return mouseOver;
	}

	/**
	 * Change the debug mode display.
	 */
	public static void changeDebugMode() {
		if (debugMode == DebugMode.NONE) {
			debugMode = DebugMode.VISIBLE;
		} else if (debugMode == DebugMode.VISIBLE) {
			debugMode = DebugMode.ALL;
		} else {
			debugMode = DebugMode.NONE;
		}
	}

	/**
	 * Get whether or not the tooltip is enabled.
	 * @return Whether or not the tooltip is enabled
	 */
	public boolean isTooltipEnabled() {
		return tooltipEnabled;
	}

	/**
	 * Set whether or not the tooltip is enabled.
	 * @param tooltipEnabled Whether or not the tooltip is enabled (true is enabled)
	 */
	public void setTooltipEnabled(boolean tooltipEnabled) {
		this.tooltipEnabled = tooltipEnabled;
		
		if (tooltipMessage == null) {
			setTooltipMessage("");
		}
	}

	/**
	 * Return the tooltip message.
	 * @return The tooltip message
	 */
	public String getTooltipMessage() {
		return tooltipMessage;
	}

	/**
	 * Set the tooltip message.
	 * @param tooltipMessage New tooltip message
	 */
	public void setTooltipMessage(String tooltipMessage) {
		this.tooltipMessage = tooltipMessage;
	}
}

File: web/src/component/Component.ts
import { Color } from '../core/Color';

export enum DebugMode {
  NONE = 'NONE',
  VISIBLE = 'VISIBLE',
  ALL = 'ALL'
}

export enum BevelType {
  NONE = 'NONE',
  IN = 'IN',
  OUT = 'OUT'
}

export enum ReferencePoint {
  TOPLEFT = 'TOPLEFT',
  TOPCENTER = 'TOPCENTER',
  TOPRIGHT = 'TOPRIGHT',
  CENTERLEFT = 'CENTERLEFT',
  CENTERCENTER = 'CENTERCENTER',
  CENTERRIGHT = 'CENTERRIGHT',
  BOTTOMLEFT = 'BOTTOMLEFT',
  BOTTOMCENTER = 'BOTTOMCENTER',
  BOTTOMRIGHT = 'BOTTOMRIGHT'
}

export interface Visible {
  isVisible(): boolean;
  setVisible(visible: boolean): void;
}

export interface Vector2f {
  x: number;
  y: number;
}

export abstract class Component implements Visible {
  private static debugMode: DebugMode = DebugMode.NONE;

  private visibleParent: Visible | null = null;
  private origin: Vector2f;
  private translation: Vector2f;
  private width: number;
  private height: number;

  private backgroundColor: Color | null = null;
  private beveled: BevelType = BevelType.NONE;
  private bevelWidth: number = 0;

  private borderColor: Color | null = null;
  private topBorderWidth: number = 0;
  private rightBorderWidth: number = 0;
  private bottomBorderWidth: number = 0;
  private leftBorderWidth: number = 0;

  protected components: Component[] = [];
  private mouseOver: boolean = false;
  private visible: boolean = true;

  private tooltipEnabled: boolean = false;
  private tooltipMessage: string = '';
  private shouldUpdateComponents: boolean = false;
  private acceptingInput: boolean = true;
  private hasFocusFlag: boolean = false;
  protected componentListeners: Array<() => void> = [];

  constructor(width: number, height: number) {
    this.origin = { x: 0, y: 0 };
    this.translation = { x: 0, y: 0 };
    this.width = width;
    this.height = height;
  }

  render(ctx: CanvasRenderingContext2D): void {
    ctx.save();
    ctx.beginPath();
    ctx.rect(this.getX(), this.getY(), this.getWidth(), this.getHeight());
    ctx.clip();

    if (this.backgroundColor !== null) {
      if (this.beveled === BevelType.NONE) {
        ctx.fillStyle = this.backgroundColor.toCSS();
        ctx.fillRect(this.getX(), this.getY(), this.getWidth(), this.getHeight());
      } else {
        const brightColor = this.backgroundColor.brighter(0.2);
        const darkColor = this.backgroundColor.darker(0.2);

        ctx.fillStyle = this.backgroundColor.toCSS();
        ctx.fillRect(
          this.getX() + this.bevelWidth,
          this.getY() + this.bevelWidth,
          this.getWidth() - this.bevelWidth * 2,
          this.getHeight() - this.bevelWidth * 2
        );

        if (this.beveled === BevelType.OUT) {
          ctx.fillStyle = brightColor.toCSS();
        } else if (this.beveled === BevelType.IN) {
          ctx.fillStyle = darkColor.toCSS();
        }
        ctx.fillRect(
          this.getX() + this.leftBorderWidth,
          this.getY() + this.topBorderWidth,
          this.getWidth() - this.leftBorderWidth - this.rightBorderWidth,
          this.bevelWidth
        );

        if (this.beveled === BevelType.OUT) {
          ctx.fillStyle = darkColor.toCSS();
        } else if (this.beveled === BevelType.IN) {
          ctx.fillStyle = brightColor.toCSS();
        }
        ctx.fillRect(
          this.getX() + this.leftBorderWidth,
          this.getY() + this.getHeight() - this.bevelWidth - this.bottomBorderWidth,
          this.getWidth() - this.leftBorderWidth - this.rightBorderWidth,
          this.bevelWidth
        );

        if (this.beveled === BevelType.OUT) {
          ctx.fillStyle = brightColor.toCSS();
        } else if (this.beveled === BevelType.IN) {
          ctx.fillStyle = darkColor.toCSS();
        }
        ctx.fillRect(
          this.getX() + this.leftBorderWidth,
          this.getY() + this.topBorderWidth,
          this.bevelWidth,
          this.getHeight() - this.topBorderWidth - this.bottomBorderWidth
        );

        if (this.beveled === BevelType.OUT) {
          ctx.fillStyle = darkColor.toCSS();
        } else if (this.beveled === BevelType.IN) {
          ctx.fillStyle = brightColor.toCSS();
        }
        ctx.fillRect(
          this.getX() + this.getWidth() - this.bevelWidth - this.rightBorderWidth,
          this.getY() + this.topBorderWidth,
          this.bevelWidth,
          this.getHeight() - this.topBorderWidth - this.bottomBorderWidth
        );
      }
    }

    for (const component of this.components) {
      component.render(ctx);

      if (
        Component.debugMode === DebugMode.ALL ||
        (Component.debugMode === DebugMode.VISIBLE && component.isVisible())
      ) {
        if (component.isVisible()) {
          ctx.strokeStyle = Color.red.toCSS();
        } else {
          ctx.strokeStyle = Color.yellow.toCSS();
        }
        ctx.strokeRect(
          component.getX(),
          component.getY(),
          component.getWidth() - 1,
          component.getHeight() - 1
        );
      }
    }

    if (this.borderColor !== null) {
      ctx.fillStyle = this.borderColor.toCSS();

      ctx.fillRect(this.getX(), this.getY(), this.getWidth(), this.topBorderWidth);

      ctx.fillRect(
        this.getX() + this.getWidth() - this.rightBorderWidth,
        this.getY(),
        this.rightBorderWidth,
        this.getHeight()
      );

      ctx.fillRect(
        this.getX(),
        this.getY() + this.getHeight() - this.bottomBorderWidth,
        this.getWidth(),
        this.bottomBorderWidth
      );

      ctx.fillRect(this.getX(), this.getY(), this.leftBorderWidth, this.getHeight());
    }

    ctx.restore();
  }

  update(delta: number): void {
    if (this.getShouldUpdateComponents()) {
      for (const c of this.components) {
        c.update(delta);
      }
    }
  }

  setShouldUpdateComponents(shouldUpdate: boolean): void {
    this.shouldUpdateComponents = shouldUpdate;
  }

  private getShouldUpdateComponents(): boolean {
    return this.shouldUpdateComponents;
  }

  setVisibleParent(visible: Visible | null): void {
    this.visibleParent = visible;
  }

  isVisible(): boolean {
    if (this.visibleParent !== null) {
      return this.visible && this.visibleParent.isVisible();
    }
    return this.visible;
  }

  setVisible(visible: boolean): void {
    this.visible = visible;
  }

  add(
    component: Component,
    location: Vector2f,
    referencePoint: ReferencePoint,
    xOffset: number = 0,
    yOffset: number = 0
  ): void {
    component.setPosition(location, referencePoint, xOffset, yOffset);
    this.components.push(component);
    component.setVisibleParent(this);
  }

  addAsColumn(
    components: Component[],
    location: Vector2f,
    xOffset: number,
    yOffset: number,
    spacing: number
  ): void {
    let currentY = location.y;
    for (const c of components) {
      this.add(c, { x: location.x, y: currentY }, ReferencePoint.TOPLEFT, xOffset, yOffset);
      currentY += c.getHeight() + spacing;
    }
  }

  addAsRow(
    components: Component[],
    location: Vector2f,
    xOffset: number,
    yOffset: number,
    spacing: number
  ): void {
    let currentX = location.x;
    for (const c of components) {
      if (c !== null) {
        this.add(c, { x: currentX, y: location.y }, ReferencePoint.TOPLEFT, xOffset, yOffset);
        currentX += c.getWidth() + spacing;
      }
    }
  }

  addAsGrid(
    components: Component[],
    location: Vector2f,
    _rows: number,
    cols: number,
    xOffset: number,
    yOffset: number,
    xSpacing: number,
    ySpacing: number
  ): void {
    let startIndex = 0;
    let currentY = location.y;

    while (startIndex < components.length) {
      const row: Component[] = [];
      for (let x = 0; x < cols && startIndex + x < components.length; x++) {
        row.push(components[startIndex + x]);
      }

      this.addAsRow(row, { x: location.x, y: currentY }, xOffset, yOffset, xSpacing);

      currentY += components[startIndex].getHeight() + ySpacing;
      startIndex += cols;
    }
  }

  addAsGrid2D(components: Component[][], xSpacing: number, ySpacing: number, location: Vector2f): void {
    let currentY = location.y;
    for (const row of components) {
      this.addAsRow(row, { x: location.x, y: currentY }, 0, 0, xSpacing);
      if (row.length > 0) {
        currentY += row[0].getHeight() + ySpacing;
      }
    }
  }

  remove(component: Component): void {
    component.setVisibleParent(null);
    const index = this.components.indexOf(component);
    if (index > -1) {
      this.components.splice(index, 1);
    }
  }

  setPosition(location: Vector2f, referencePoint: ReferencePoint, xOffset: number = 0, yOffset: number = 0): void {
    const offsetLocation: Vector2f = { x: location.x + xOffset, y: location.y + yOffset };
    const x = offsetLocation.x;
    const y = offsetLocation.y;

    switch (referencePoint) {
      case ReferencePoint.TOPLEFT:
        this.setLocation(x, y);
        break;
      case ReferencePoint.TOPCENTER:
        this.setLocation(x - this.getWidth() / 2, y);
        break;
      case ReferencePoint.TOPRIGHT:
        this.setLocation(x - this.getWidth(), y);
        break;
      case ReferencePoint.CENTERLEFT:
        this.setLocation(x, y - this.getHeight() / 2);
        break;
      case ReferencePoint.CENTERCENTER:
        this.setLocation(x - this.getWidth() / 2, y - this.getHeight() / 2);
        break;
      case ReferencePoint.CENTERRIGHT:
        this.setLocation(x - this.getWidth(), y - this.getHeight() / 2);
        break;
      case ReferencePoint.BOTTOMLEFT:
        this.setLocation(x, y - this.getHeight());
        break;
      case ReferencePoint.BOTTOMCENTER:
        this.setLocation(x - this.getWidth() / 2, y - this.getHeight());
        break;
      case ReferencePoint.BOTTOMRIGHT:
        this.setLocation(x - this.getWidth(), y - this.getHeight());
        break;
    }
  }

  getPosition(referencePoint: ReferencePoint): Vector2f {
    const x = this.getX();
    const y = this.getY();

    switch (referencePoint) {
      case ReferencePoint.TOPLEFT:
        return { x, y };
      case ReferencePoint.TOPCENTER:
        return { x: x + this.getWidth() / 2, y };
      case ReferencePoint.TOPRIGHT:
        return { x: x + this.getWidth(), y };
      case ReferencePoint.CENTERLEFT:
        return { x, y: y + this.getHeight() / 2 };
      case ReferencePoint.CENTERCENTER:
        return { x: x + this.getWidth() / 2, y: y + this.getHeight() / 2 };
      case ReferencePoint.CENTERRIGHT:
        return { x: x + this.getWidth(), y: y + this.getHeight() / 2 };
      case ReferencePoint.BOTTOMLEFT:
        return { x, y: y + this.getHeight() };
      case ReferencePoint.BOTTOMCENTER:
        return { x: x + this.getWidth() / 2, y: y + this.getHeight() };
      case ReferencePoint.BOTTOMRIGHT:
        return { x: x + this.getWidth(), y: y + this.getHeight() };
    }
  }

  getBackgroundColor(): Color | null {
    return this.backgroundColor;
  }

  setBackgroundColor(color: Color | null): void {
    this.backgroundColor = color;
  }

  setBevel(bevelType: BevelType): void {
    this.beveled = bevelType;
  }

  setBevelWidth(width: number): void {
    this.bevelWidth = width;
  }

  setBorderWidth(width: number): void {
    this.setTopBorderWidth(width);
    this.setRightBorderWidth(width);
    this.setBottomBorderWidth(width);
    this.setLeftBorderWidth(width);
  }

  setTopBorderWidth(width: number): void {
    this.topBorderWidth = width;
  }

  setRightBorderWidth(width: number): void {
    this.rightBorderWidth = width;
  }

  setBottomBorderWidth(width: number): void {
    this.bottomBorderWidth = width;
  }

  setLeftBorderWidth(width: number): void {
    this.leftBorderWidth = width;
  }

  setBorderColor(color: Color | null): void {
    this.borderColor = color;
  }

  containsPoint(x: number, y: number): boolean {
    return (
      x >= this.getX() &&
      x <= this.getX() + this.getWidth() &&
      y >= this.getY() &&
      y <= this.getY() + this.getHeight()
    );
  }

  getWidth(): number {
    return this.width;
  }

  getHeight(): number {
    return this.height;
  }

  getX(): number {
    return this.origin.x + this.translation.x;
  }

  getY(): number {
    return this.origin.y + this.translation.y;
  }

  setLocation(x: number, y: number): void {
    if (this.components.length > 0) {
      const deltaX = x - this.getX();
      const deltaY = y - this.getY();
      for (const c of this.components) {
        c.setLocation(c.getX() + deltaX, c.getY() + deltaY);
      }
    }
    this.origin.x = x;
    this.origin.y = y;
  }

  setTranslation(x: number, y: number): void {
    if (this.components.length > 0) {
      for (const c of this.components) {
        c.setTranslation(x, y);
      }
    }
    this.translation.x = x;
    this.translation.y = y;
  }

  setAcceptingInput(acceptingInput: boolean): void {
    this.acceptingInput = acceptingInput;

    if (!acceptingInput) {
      this.mouseOver = false;
    }

    for (const c of this.components) {
      c.setAcceptingInput(acceptingInput);
    }
  }

  isAcceptingInput(): boolean {
    return this.acceptingInput;
  }

  getVisibleParent(): Visible | null {
    return this.visibleParent;
  }

  mouseMoved(_oldx: number, _oldy: number, newx: number, newy: number): void {
    if (!this.isVisible() || !this.isAcceptingInput()) {
      return;
    }

    this.mouseOver = this.containsPoint(newx, newy);

    if (this.isMouseOver() && this.tooltipEnabled) {
      // TODO: Scene.showTooltip(newx, newy, this, tooltipMessage);
    }
  }

  mousePressed(_button: number, mx: number, my: number): void {
    if (!this.isVisible() || !this.isAcceptingInput()) {
      return;
    }

    this.mouseOver = this.containsPoint(mx, my);
  }

  mouseReleased(_button: number, mx: number, my: number): void {
    if (!this.isVisible() || !this.isAcceptingInput()) {
      return;
    }

    this.mouseOver = this.containsPoint(mx, my);
  }

  isMouseOver(): boolean {
    return this.mouseOver;
  }

  static changeDebugMode(): void {
    if (Component.debugMode === DebugMode.NONE) {
      Component.debugMode = DebugMode.VISIBLE;
    } else if (Component.debugMode === DebugMode.VISIBLE) {
      Component.debugMode = DebugMode.ALL;
    } else {
      Component.debugMode = DebugMode.NONE;
    }
  }

  isTooltipEnabled(): boolean {
    return this.tooltipEnabled;
  }

  setTooltipEnabled(tooltipEnabled: boolean): void {
    this.tooltipEnabled = tooltipEnabled;

    if (this.tooltipMessage === '') {
      this.setTooltipMessage('');
    }
  }

  getTooltipMessage(): string {
    return this.tooltipMessage;
  }

  setTooltipMessage(tooltipMessage: string): void {
    this.tooltipMessage = tooltipMessage;
  }

  hasFocus(): boolean {
    return this.hasFocusFlag;
  }

  setFocus(focus: boolean): void {
    this.hasFocusFlag = focus;
  }

  addListener(listener: () => void): void {
    this.componentListeners.push(listener);
  }

  removeListener(listener: () => void): void {
    const index = this.componentListeners.indexOf(listener);
    if (index > -1) {
      this.componentListeners.splice(index, 1);
    }
  }

  protected notifyListeners(): void {
    for (const listener of this.componentListeners) {
      listener();
    }
  }

  keyReleased(_key: string, _char: string): void {
    // Override in subclasses
  }
}
